libuavcan/include/libuavcan/platform/memory.hpp

```diff
@@ -0,0 +1,313 @@
+        for (std::size_t i = NumBlocksParam - 1; i > 0; --i)
+        {
+            Block* previous = &list[i - 1];
+            previous->next  = next;
+            next            = previous;
+        }
+    }
+
+    // +------------------------------------------------------------------+
+    // | RULE OF SIX
+    // +------------------------------------------------------------------+
+    ~StaticMemoryPool() = default;
+
+public:
+    StaticMemoryPool(const StaticMemoryPool&)  = delete;
+    StaticMemoryPool(const StaticMemoryPool&&) = delete;
+    StaticMemoryPool& operator=(const StaticMemoryPool&) = delete;
+    StaticMemoryPool& operator=(const StaticMemoryPool&&) = delete;
+
+private:
+    // +------------------------------------------------------------------+
+    // | ALLOCATORS
+    // |    A list of allocators that can use this memory pool.
+    // +------------------------------------------------------------------+
+    template <std::size_t, std::uint8_t, typename, typename>
+    friend class PoolAllocator;
+
+    /**
+     * Get a reference to the static memory pool. The first time this is called
+     * the memory pool will perform internal initialization that is only thread-safe
+     * if the compiler has generated the proper thread-safe static access instructions.
+     */
+    static StaticMemoryPool& getReference()
+    {
+        static StaticMemoryPool pool_;
+        return pool_;
+    }
+
+    void* allocate(std::size_t size)
+    {
+        Block* previous = nullptr;
+        if (size <= BlockSizeParam)
+        {
+            Block* next = nullptr;
+            do
+            {
+                previous = free_list_.load();
+                if (previous == nullptr)
+                {
+#ifdef LIBUAVCAN_ENABLE_EXCEPTIONS
+                    std::bad_alloc exception;
+                    throw exception;
+#endif
+                    break;
+                }
+                next = previous->next;
+            } while (!std::atomic_compare_exchange_weak(&free_list_, &previous, next));
+        }
+        else
+        {
+#ifdef LIBUAVCAN_ENABLE_EXCEPTIONS
+            std::bad_alloc exception;
+            throw exception;
+#endif
+        }
+
+        return previous;
+    }
+
+    void deallocate(void* ptr)
+    {
+        if (ptr != nullptr)
+        {
+            Block* reclaimed_block = reinterpret_cast<Block*>(ptr);
+            reclaimed_block->next  = std::atomic_exchange(&free_list_, reclaimed_block);
+        }
+    }
+};
+
+/**
+ * Adapter to provide STL and other instances access to a shared memory pool.
+ *
+ * @tparam  NumBlocksParam      The number of blocks to allocate in the memory pool.
+ * @tparam  BlockSizeParam      The size in bytes of each block in the memory pool.
+ * @tparam T                    A type alias used to pretend that the pointers returned are pointers to
+ *                              this type. By default this is void which means void* is returned like
```

Comment:
> By default this is void which means void* is returned

Looks like uint8 to me.

Reply:
Yeah, I changed the default but missed the comment. Good catch.